enhancements/machine-api/aws-placement-groups.md

```diff
@@ -0,0 +1,538 @@
+
+As group names must be unique within a region, we should not fall foul of eventual consistency issues leading to
+leaking of placement groups as we can do with EC2 instances.
+
+#### Removal of Placement Groups
+
+To ensure we leave user accounts clean, we should remove any placement group that is no longer required.
+
+Each time a Machine is removed, we must check whether it belongs to a placement group.
+If it belongs to a placement group, we must also check that there are no other instances within the placement group.
+This will mean listing all instances in the region, filtering them down based on their placement group name.
+
+We should be able to replicate the following CLI command within the AWS Machine controller for this purpose:
+```shell
+aws ec2 describe-instances --filters Name=placement-group-name,Values=<placement-group-name>
+```
+
+If no other instance exists within the group, we will also delete the placement group once the instance has been
+removed.
+
+Before deleting the placement group, the controller must ensure that the placement group was owned by the cluster.
+The tags on the placement group should be inspected to ensure they match the expected format.
+We should also check whether, if required, we will be able to recreate the placement group at a later date.
+As the `groupType` is going to be required to use placement groups, we should be able to create a new
+placement group if we remove it.
+For `partition` groups, we will also need to validate the partition count is valid before we remove the
+placement group.
+
+If the placement group includes the standard cluster ownership tag, and the Machine has the `groupType` set, it will be
+pruned as described above.
+
+#### General limitations of Placement Groups
+
+- You can create a maximum of 500 placement groups per account in each Region.
+- The name that you specify for a placement group must be unique within your AWS account for the Region.
+- An instance can be launched in one placement group at a time; it cannot span multiple placement groups.
+- You cannot launch AWS Dedicated Hosts in placement groups.
+
+#### Installer changes
+
+We do not plan to add Placement group configuration to the installer machine pool configuration.
+Should users wish to launch the day 0 worker machines within a placement group, they will be able to modify the
+installer generated machineset manifests before the cluster is created.
+We do not intend to support placement groups for control plane machines at all at this point.
+
+We will however need to update the installer cluster deprovisioning logic, as currently it does not know how to remove
+a placement group if one is present within the cluster when the destroy operation is started.
+
+#### Additional permissions
+
+As this enhancement introduces new API calls to the Machine API, we will need to introduce 2 new permissions within the
+CredentialsRequest.
+
+```yaml
+- ec2:DescribePlacementGroups
+- ec2:CreatePlacementGroup
```

Comment:
ec2:DeletePlacementGroup, too, right?

If so, then this permission also needs to be added to the permissions checked by the installer since it will be needed by the destroyer.

Reply:
Good catch, I hadn't done any testing of the deletion (or implemented it) when this was written so it slipped through. It is needed though so will add a note.

Do you happen to have a link to roughly where the list of permissions the installer checks is sourced from? Can raise a PR

Reply:
https://github.com/openshift/installer/blob/e9bd965ebc73bf2fd8c2b052ecf3918437727c0a/pkg/asset/installconfig/aws/permissions.go#L175-L192

Reply:
PR https://github.com/openshift/installer/pull/5655